Doc/sam/event.3
.de F
.B
.if !"\\$1"" \&\\$1 \\$2 \\$3 \\$4 \\$5 \\$6
..
.de L
.B
.if !"\\$1"" \&\\$1 \\$2 \\$3 \\$4 \\$5 \\$6
..
.de FR
.BR "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de LR
.BR "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de CW
.ft B
..
.\" This is gross but it avoids relying on internal implementation details
.\" of the -man macros.
.de TF
.IP "" \w'\fB\\$1\ \ \fP'u
.PD0
..
.de EX
.CW
.nf
..
.de EE
.fi
..
.\" delete above this point if your system has F, L, FR, LR, CW and TF macros
.TH EVENT 3G
.SH NAME
event, einit, estart, etimer, eread, emouse, ekbd, ecanread, ecanmouse, ecankbd, ereshaped, getrect, menuhit, Event, Mouse, Menu \- graphics events
.SH SYNOPSIS
.nf
.B
#include <libc.h>
.B
#include <libg.h>
.PP
.ta \w'\fLunsigned long 'u
.PP
.B
void	einit(ulong keys)
.PP
.B
unsigned long	event(Event *e)
.PP
.B
Mouse	emouse(void)
.PP
.B
int	ekbd(void)
.PP
.B
int	ecanmouse(void)
.PP
.B
int	ecankbd(void)
.PP
.B
unsigned long	estart(ulong key, int fd, int n)
.PP
.B
unsigned long	etimer(ulong key, int n)
.PP
.B
unsigned long	eread(ulong keys, Event *e)
.PP
.B
int	ecanread(ulong keys)
.PP
.B
void	ereshaped(Rectangle r)
.PP
.B
Rectangle	getrect(int but, Mouse *m)
.PP
.B
int	menuhit(int but, Mouse *m, Menu *menu)
.PP
.ft L
enum{
	Emouse = 1,
	Ekeyboard = 2,
}
.ft P
.fi
.SH DESCRIPTION
These routines provide an interface to multiple sources of input.
To use them,
.I einit
must be called.
If the argument to
.I enit
has the
.B Emouse
and
.B Ekeyboard
bits set,
the mouse and keyboard events will be enabled;
in this case,
.IR xtbinit
(see
.IR graphics (3))
must have already been called.
The user must provide a function called
.IR ereshaped ,
which will be called whenever the window in which the process
is running has been reshaped; the argument will be the Rectangle
for the new window shape, including the border.
.PP
As characters are typed on the keyboard, they are read by the
event mechanism and put in a queue.
.I Ekbd
returns the next character from the queue, blocking until the
queue is non-empty.
The characters are read by the event mechanism from the keyboard
so they are available as soon as they are typed.
.PP
When the mouse moves or a mouse button is depressed or released,
a new mouse event is queued by the event mechanism.
.I Emouse
returns the next mouse event from the queue, blocking until the
queue is non-empty.
.I Emouse
returns a
.B Mouse
structure:
.IP
.EX
.ta 6n +\w'unsigned long 'u
struct Mouse
{
	int	buttons;
	Point	xy;
	unsigned long msec;
};
.EE
.PP
.B Buttons
is a bit field;
.B buttons&1
is set when the left mouse button is depressed,
.B buttons&2
when the middle button is depressed,
and
.B buttons&4
when the right button is depressed.
The current mouse position is always returned in
.BR xy .
.BR Msec
is the time stamp of the mouse event in units of milliseconds.
.PP
.I Ecankbd
and
.I ecanmouse
return non-zero when there are keyboard or mouse events to available
to be read.
.PP
.I Estart
can be used to register additional file descriptors.
It takes as arguments the file descriptor to register,
the maximum length of an event message on that descriptor,
and a key to be used in accessing the event.
The key must be a power of 2 and must not confilict with any previous keys.
If a zero key is given, one which is not used will be chosen and returned.
.B
Ekeyboard
and
.B Emouse
are the mouse and keyboard event keys.
.PP
.I Etimer
starts a repeating timer with a period of n milliseconds (default 1 second).
Only one timer can be started.
Extra timer events are not queued and the timer channel has no associated data.
.PP
.I Eread
waits for the next event specified by the mask
.B keys
of event keys submitted to estart.
It fills in the appropriate field of the argument
.B Event
structure, which looks like:
.IP
.EX
struct Event
{
	int	kbdc;
	Mouse	mouse;
	int	n;
	uchar	data[EMAXMSG];
}
.EE
.PP
.B Data
is an array which is large enough to hold a plan 9 protocol message.
.I Eread
returns the key for the event which was chosen.
For example, if a mouse event was read,
.I Emouse
will be returned.
.PP
.I Event
waits for the next event of any kind.
The return is the same as for
.IR eread .
.PP
As described in
.IR graphics (3),
the graphics functions are buffered.
.IR Event ,
.IR eread ,
.IR emouse ,
and
.I ekbd
all cause a buffer flush unless there is an event of the
appropriate type ready to return.
.PP
.I Getrect
is used to prompt the user to sweep a rectangle.
It should be called with
.I m
holding the mouse event that triggered the
.I getrect
(or, if none, a
.B Mouse
with
.B buttons
set to 7).
It changes to the sweep cursor,
waits for the buttons to all go up,
and then waits for button number
.I but
to be depressed, marking the initial corner.
If another button is depressed instead,
.I getrect
returns a rectangle
with zero for both x-coordinates, after
waiting for all the buttons to be released.
Otherwise,
.I getrect
continually draws the swept rectangle
until the button is released again, and returns the swept rectangle.
The mouse structure pointed to by
.I m
will contain the final mouse event.
.PP
.I Menuhit
displays a menu and returns a seleced menu item number.
It should be called with
.I m
holding the mouse event that triggered the
.I menuhit .
A
.B Menu
is a structure:
.IP
.EX
struct Menu
{
	char	**item;
	char	*(*gen)(int);
	int	lasthit;
}
.EE
.PP
If
.B item
is nonzero, it should be a null-terminated array of the character strings
to be displayed as menu items.
Otherwise,
.B gen
should be a function that, given an item number, returns the character
string for that item, or zero if the number is past the end of the list.
Items are numbered starting at zero.
.I Menuhit
waits until
.I but
is released, and then returns the number of the selection,
or \(mi1 for no selection.
The
.I m
argument is filled in with the final mouse event.
.SH "SEE ALSO"
.IR graphics (3).


Doc/changes.txt
0.13.12 started this file
* Each line in $WILYTOOLS or $HOME/.wilytools can contain 
either a comment (starts with whitespace or '#') or a regexp,
then a tab, then some tools (i.e. some text) which will be put
into the tag of any window whose label matches the regexp.
The _first_ regexp match wins.
0.13.14
* Scrolling forward works better
* plugged some memory leaks
0.13.15 
fixed silly bug in text_read: Axel Belinfant 
patch from Steve Kotsopoulos - $WMAINTAG and $WCOLTAG 
removed last_escape_select 
newline only special in tag now 
h version 
Version builtin removed 
no longer warps cursor when expanding tile after typing in it 
utfwrite handles short writes -- Bill Trost 
wilytools have newline stripped, b2b1 on Anchor selects line 
changed mouse_expand: clicking _near_ a selection hits the selection 
use view_cut a lot 
removed some functions in keyboard.c 
use view->autoindent to decide to autoindent a view 
Indent builtin 
option -a to set tabsize 
0.13.16 
bugfix with B2B1 
removed Version from default wilytools 
word-delete seems saner to me 
B3 checks for local files before include files 
0.13.17 
splitting data.c up 
tidied a few functions in data.c 
cloned windows share the same tag, modifying label works better 
tag modification seems to be working 
0.13.18 
ctrlu patch -- Bill 
removed some dead code from mouse.c 
renamed globals.h global.h 
updated MANIFEST 
moved columnate to from dir.c to util.c 
moved more out of dir.c 
removed a few syntax errors 
broken view.c into different files 
renamed view_left view_hscroll 
moved most of selection_fd to text_fd 
more from view.c to util.c 
removed selection_fd altogether 
Kill patch from Steve_Kilbane 
replaced view_isvisible with ISVISIBLE 
compiles with view.c all split up 
renamed the view-related files 
renamed the view-related files 
0.13.19 
scrollbugfix from Steve_Kilbane 
removed text.h from view.c 
tinkering with vgeom 
moved stuff from vsearch to text2 
moved b3 from mouse.c to vsearch.c 
view_goto simplified 
lotsa internal changes 
rename data_labelmodified to data_setlabel 
rename data_backall to data_backupall 
broke search.c into click.c and sambuf.c 
removed text_clear 
removed text_delete 
split and tidied text2.c 
uses sam regexp 
0.13.20 
remove garyfind hack
bugfix undo.c (was causing core dumps)
msg_new now matches doco
0.13.21
exec dies if cannot chdir 
tile patch from Bill Trost
0.13.22
utfwrite handles write returning
_all_ the resize boxes of a window are updated when the window is written
right undo stuff happens for directory read
misc. patches from Bill Trost
realpath uses cached value of cwd
tag_init no longer uses fgets
0.13.23

selected window display patch -- Bill Trost 
directory speedup patch -- Bill Trost 
aesthetic improvements patch -- awatson@NMSU.Edu 
made Alan's and Bill's aesthetics patches cooperate a bit 
noutput bugfix -- Bill Trost, Alan Watson 
0.13.24

make sure anchor is at end of tag for new windows 
organising builtins
bugfix: scrollbar geometry a bit wierd
separated out all operations on lines into line.c 
scrolling backwards through a file now much faster
text_linenumber simplified 
tute/start updated a bit
0.13.25
stop complaining if cannot open .wilytools -- awatson@NMSU.Edu
prettier scrolling, wily -e -- awatson@NMSU.Edu 
text_replace in tag doesn't get forwarded 
fix the problems with switching between Fonts -- Bill Trost 
append WMAINTAG, rather than replace it -- Steve Kotsopoulos 
undo patch --Bill
fixes to above --  Gary 
getname, gettools -- gary
0.13.26


wily/event.c
/*********************************************************
	Set up and handle interactions with all the libXg
	events that aren't mouse,timer or keyboard.
*********************************************************/

#include "wily.h"
#include <errno.h>
#include <signal.h>

typedef struct Key Key;

/* Different uses for a libXg event key */
typedef enum Keytype {
	Kfree, 	/* not in use */
	Klisten, 	/* connection requests */
	Kout, 	/* output to stderr or a window */
	Kmsg	/* from external connections */
} Keytype; 

/* Info about a libXg event key */
struct Key {
	Keytype 	t;
	int		fd;		/* file descriptor*/
	ulong	key;		/* libXg event key */
	
	/* 
	 * Buffer for incomplete messages.
	 * Only in use if k->t == Kmsg.
	 */
	Mbuf	buf;		

	/*
	 * Info about the program sending us output.
	 * Only valid if k->t == Kout.
	 */
	int	pid;
	Path	cmd;	/* shortened name of executing command */
	Path	olabel;	/* where output should go */

	/* If output should replace some View's selection ( |, > ).
	 * Note that when we use |cmd or >cmd, we'll get TWO
	 * event Keys, one for stderr (with k->v == 0), and
	 * one for stdout, (with k->v set).
	 */
	View		*v;
	Bool		first;

	/*
	** Runes with multibyte UTF codes must not be split across
	** block boundaries.  Strip them off and join them in this buffer.
	*/
	char		obuf[UTFmax+1];
	int		nobuf;
};

/*
 * Because libXg can only keep track of a small number of
 * event keys, so do we.  We keep the Keys in a static array.
 * The Key at position 'n' in keytab has libXg event key 2^n.
 *
 * This is a bit wasteful, and a bit inefficient.  Some event keys
 * will never be used here.  Too bad.
 */
enum {
	MAXKEYS =FOPEN_MAX
};
static Key		keytab[MAXKEYS];

static Key*	key_find(ulong key);
static void	outmsg(Key*k, int n, char*s);
static void	key_del(Key *k);
static Key*	key_new(ulong key, int fd, Keytype t);
static Key*	key_findcmd(char*cmd);
static void	oneword(char*s, char*l, int s_length);
static void	output_fullutfs(Key*k, int n, char*s);
static void	output(Key*k, int n, char*s);
static void	ex_accept(int n, char*s);

/* Initialise keytab */
void
keytab_init(void) {
	Key	*k;
	ulong	key = 1;

	for(k = keytab; k < keytab + MAXKEYS; k++) {
		k->t = Kfree;
		k->key = key;
		key *= 2;
		mbuf_init(&k->buf);
	}
}

/*
 * Start listening for output on 'fd'.
 * The output is from the (long) command 'cmd', process 'pid'.
 * The context is associated with 'label'.
 * If 'v' is set, we are piping the output to this View.
 */
int
event_outputstart(int fd, int pid, char*cmd, char*label, View *v){
	ulong key;
	Key	*k;
	
	if (!(key = estart(0, fd, 0))) {
		diag(0, "estart");
		return -1;
	}
	k = key_new(key, fd, Kout);
	k->pid = pid;
	k->v = v;

	/* Every 'pipe' key has another 'stderr' key, we only
	 * display the 'cmd' associated with the 'stderr' key.
	 */
	if(k->v) {
		k->first = true;
	} else {
		oneword(k->cmd, cmd, sizeof(k->cmd));
		olabel(k->olabel, label);
		addrunning(k->cmd);
	}

	k->nobuf = 0;

	return 0;
}

/* Start listening for connection attempts on 'fd' */
void
event_wellknown(int fd){
	ulong key;
	
	if(!(key = estart(0, fd, 0)))
		error("estart");
	key_new(key,fd, Klisten);
}

/* 
 * Handle 'n' bytes of data in 's' which arrived from 
 * libXg event key 'key'.
 */
void
dofd(ulong key, int n, char*s) {
	Key	*k = key_find(key);

	if(!n) {
		key_del(k);
		return;
	}

	switch(k->t){
	case Klisten:	ex_accept(n,s);	break;
	case Kout:	output_fullutfs(k, n, s);	break;
	case Kmsg:	outmsg(k, n, s);	break;
	default:		error("bad key type %d", k->t);		break;
	}
}

/*
 * Kill all the external processes matching
 * the arguments.
 */
void
kill_all(char*s) {
	Key	*k = 0;
	const char *sep = " \t\n";

	for (s = strtok(s, sep); s; s = strtok(NULL, sep)) {
		if ((k = key_findcmd(s))) {
			/* only output keys have a prog attached */
			assert(k->t == Kout); 
			if(kill(- k->pid, SIGKILL))
				diag(0, "kill %s", s);
		}
	}
}

/* Output a list of possible 'Kill' commands. */
void
kill_list(void){
	Key	*k;
	
	errno = 0;
	for(k = keytab; k < keytab + MAXKEYS; k++)
		if(k->t == Kout && !k->v)
			diag(0, "Kill %s", k->cmd);
}

/******************************************************
	static functions
******************************************************/
static void
outmsg(Key*k, int n, char*s) {
	if(partialmsg(&k->buf, k->fd, n, s)){
		diag(0, "Received bad message, closing connection");
		key_del(k);
	}
}

/* Free any resources that 'k' was tying up. */
static void
key_del(Key *k) {
	close(k->fd);
	estop(k->key);

	switch(k->t) {
	case Kout: 	if (k->nobuf > 0)	output(k, k->nobuf, k->obuf);
				if (!k->v)			rmrunning(k->cmd);
				break;
	case Kmsg:	data_fdstop(k->fd); break;
	case Klisten:	error("Klisten closed");
	default:		error("bad key type %d", k->t);
	}
	k->t = Kfree;
}

/* Allocate and return a new Key with the given key, fd and type */
static Key*
key_new(ulong key, int fd, Keytype t) {
	Key	*k;

	k = key_find(key);
	assert(k->t == Kfree);
	k->fd = fd;
	k->t = t;
	return k;
}

/* Find the output Key with the given 'cmd' */
static Key*
key_findcmd(char*cmd)
{
	Key	*k;

	for(k = keytab; k < keytab + MAXKEYS; k++)
		if (k->t == Kout && STRSAME(cmd, k->cmd))
			return k;
	return 0;
}

/* Put a one-word version of long command 'l' into 's' */
static void
oneword(char*s, char*l, int s_length){
	char *blank = strpbrk(l, whitespace);
	if ((blank) && (blank - l < s_length))
		s_length = blank - l + 1;
	strncpy(s, l, s_length);
	s[s_length-1] = '\0'; /* some strncpy() don't do this */
}

/*
** Key has a buffer where we concatenate runes that get split between two
** (or more -- very small?) blocks.  This is accomplished in four steps:
** 1)  If we have an uncomplete rune from last call, try to complete it.
** 2)  If output isn't exhausted and we have something from last call,
** output it -- it's either a full rune or bad utf.
** 3)  Put away any unfull rune at the and of the output.
** 4)  Output any remaining output, which is now only full utfs.
*/
static void
output_fullutfs(Key *k, int n, char *s) {
	if (k->nobuf > 0) {
		int m = fillutfchar(k->obuf, k->nobuf, s, n);
		k->nobuf+=m, s+=m, n-=m;
	}
	if ((n > 0) && (k->nobuf > 0)) {
		output(k, k->nobuf, k->obuf);
		k->nobuf = 0;
	}
	if (n > 0) {
		int m = unfullutfbytes(s, n);
		if (m > 0) {
			memcpy(k->obuf, s+n-m, m);
			k->obuf[m] = '\0';
			n -= m;
		}
		k->nobuf=m;
	}
	if (n > 0) {
		output(k, n, s);
	}
}

/* Copy the 'n' bytes of data in 's' to the window or output directory
 * corresponding to 'k'
 */
static void
output(Key*k, int n, char*s)  {
	if(k->v)
		view_pipe(k->v, &k->first, s, n);
	else
		noutput(k->olabel, s, n);
}

/* Accept the connection we're told about with 's' and 'n' */
static void
ex_accept(int n, char*s) {
	int	fd;
	ulong	key;
	Key		*k;

	if((fd =wily_connect(s,n))<0) {
		diag(0, "failed connection attempt");
		return;
	}
	
	if (!(key= estart(0,fd,0))){
		diag(0, "couldn't estart to accept connection");
		close(fd);
		return;
	}
	k = key_new(key, fd, Kmsg);
	mbuf_clear(&k->buf);
}

/*
 * Find the Key info given the libXg event key.
 * 
 * ASSUMPTION: integers at least 32 bit.
 */
static Key*
key_find(ulong key) {
	Key	*k;

	for(k = keytab; k < keytab + MAXKEYS; k++)
		if(k->key == key)
			return k;
	
	/* We assume key_find will always work */
	assert(false);	
	return 0;
}

